server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -126,7 +128,17 @@ private Map<String, UploadedSegmentMetadata> readLatestMetadataFile() throws IOE
 
     private Map<String, UploadedSegmentMetadata> readMetadataFile(String metadataFilename) throws IOException {
         try (IndexInput indexInput = remoteMetadataDirectory.openInput(metadataFilename, IOContext.DEFAULT)) {
-            Map<String, String> segmentMetadata = indexInput.readMapOfStrings();
+            // BufferedChecksumIndexInput reader keeps computing checksum for all the bytes we read from indexInput
+            // This checksum is then compared with checksum in footer
+            ChecksumIndexInput in = new BufferedChecksumIndexInput(indexInput);
+            CodecUtil.checkHeader(
+                in,
+                UploadedSegmentMetadata.METADATA_CODEC,
+                UploadedSegmentMetadata.CURRENT_VERSION,
+                UploadedSegmentMetadata.CURRENT_VERSION
+            );
+            Map<String, String> segmentMetadata = in.readMapOfStrings();
```

Comment:
This similarly can be abstracted so that we have history of encoding and decoding the DataInput or DataOutput.